pyccl/halos/profiles.py

```diff
@@ -0,0 +1,824 @@
+            mass_def (:obj:`MassDef`): a mass definition object.
+
+        Returns:
+            float or array_like: halo profile. The shape of the
+            output will be `(N_M, N_r)` where `N_r` and `N_m` are
+            the sizes of `r` and `M` respectively.
+        """
+        if getattr(self, '_cumul2d', None):
+            s_r_t = self._cumul2d(cosmo, r_t, M, a, mass_def)
+        else:
+            s_r_t = self._projected_fftlog_wrap(cosmo, r_t, M,
+                                                a, mass_def,
+                                                is_cumul2d=True)
+        return s_r_t
+
+    def _fftlog_wrap(self, cosmo, k, M, a, mass_def,
+                     fourier_out=False,
+                     large_padding=True):
+        # This computes the 3D Hankel transform
+        #  \rho(k) = 4\pi \int dr r^2 \rho(r) j_0(k r)
+        # if fourier_out == False, and
+        #  \rho(r) = \frac{1}{2\pi^2} \int dk k^2 \rho(k) j_0(k r)
+        # otherwise.
+
+        # Select which profile should be the input
+        if fourier_out:
+            p_func = self._real
+        else:
+            p_func = self._fourier
+        k_use = np.atleast_1d(k)
+        M_use = np.atleast_1d(M)
+        lk_use = np.log(k_use)
+        nM = len(M_use)
+
+        # k/r ranges to be used with FFTLog and its sampling.
+        if large_padding:
+            k_min = self.precision_fftlog['padding_lo_fftlog'] * np.amin(k_use)
+            k_max = self.precision_fftlog['padding_hi_fftlog'] * np.amax(k_use)
+        else:
+            k_min = self.precision_fftlog['padding_lo_extra'] * np.amin(k_use)
+            k_max = self.precision_fftlog['padding_hi_extra'] * np.amax(k_use)
+        n_k = (int(np.log10(k_max / k_min)) *
+               self.precision_fftlog['n_per_decade'])
+        r_arr = np.geomspace(k_min, k_max, n_k)
+
+        p_k_out = np.zeros([nM, k_use.size])
+        # Compute real profile values
+        p_real_M = p_func(cosmo, r_arr, M_use, a, mass_def)
+        # Power-law index to pass to FFTLog.
+        plaw_index = self._get_plaw_fourier(cosmo, a)
+
+        # Compute Fourier profile through fftlog
+        k_arr, p_fourier_M = _fftlog_transform(r_arr, p_real_M,
+                                               3, 0, plaw_index)
+        lk_arr = np.log(k_arr)
+
+        for im, p_k_arr in enumerate(p_fourier_M):
+            # Resample into input k values
+            p_fourier = resample_array(lk_arr, p_k_arr, lk_use,
+                                       self.precision_fftlog['extrapol'],
+                                       self.precision_fftlog['extrapol'],
+                                       0, 0)
+            p_k_out[im, :] = p_fourier
+        if fourier_out:
+            p_k_out *= (2 * np.pi)**3
+
+        if np.ndim(k) == 0:
+            p_k_out = np.squeeze(p_k_out, axis=-1)
+        if np.ndim(M) == 0:
+            p_k_out = np.squeeze(p_k_out, axis=0)
+        return p_k_out
```

Comment:
The way you squeeze the outputs here doesn't match the doc string.

Reply:
Humm... The docs say
> The shape of the output will be `(N_M, N_k)` where `N_k` and `N_m` are the sizes of `k` and `M` respectively.
so I think this is correct, right?

Reply:
Well if N_M = 1, then I think you will squeeze that dimension out.

Reply:
Oh, so you mean that I should specify that the dimensions get squeezed out if you pass a scalar?

Reply:
Yeah for sure. These shape errors can be subtle and silent.

Reply:
Done